File: Remove and Reverse - GFG/remove-and-reverse.java
//{ Driver Code Starts
//Initial Template for Java
import java.io.*;
import java.util.*; 
class GFG{
    public static void main(String args[]) throws IOException { 
        BufferedReader read = new BufferedReader(new InputStreamReader(System.in));
        int t = Integer.parseInt(read.readLine());
        
        while(t-- > 0){
          
            String S = read.readLine().trim();
            Solution ob = new Solution();
            String ans = ob.removeReverse(S); 
            System.out.println(ans);
        }
    } 
} 

// } Driver Code Ends
//User function Template for Java

class Solution 
{ 
    String removeReverse(String S) 
    { 
        // code here
          StringBuffer s = new StringBuffer(S);
        int freq[] = new int[26];
        for(int i=0;i<s.length();i++){
            char ch = s.charAt(i);
            freq[ch-'a']++;
        }
        //logic
        int flag=0;
        int st=0; int end=s.length()-1; 
        while(st<=end)
        {
            if(flag==0)
            {
                if(freq[s.charAt(st)-'a']==1) ///no reperatinng ch , check for others
                st++;
                
                else{
                    freq[s.charAt(st)-'a']--;
                    s.replace(st,st+1,"#");
                    st++;
                    flag=1; // string is reverse , traverse from 'end' ptr for f==1
                }
                
            }
            else
            {
                if(freq[s.charAt(end)-'a']==1) ///no reperatinng ch , check for others
                end--;
                
                else{
                    freq[s.charAt(end)-'a']--;
                    s.replace(end,end+1,"#");
                    end--;
                    flag=0;
                }
            }
        }
         if(flag==1)
            s.reverse();
        
        //
         String ans="";
        for(int i=0;i<s.length();i++)
        {
            if(s.charAt(i)!='#')
                ans+=s.charAt(i);
        }
        return ans;
        
    }
} 

//{ Driver Code Starts.

// } Driver Code Ends

File: Remove and Reverse - GFG/remove-and-reverse.cpp
//{ Driver Code Starts
//Initial Template for C++
#include <bits/stdc++.h>
using namespace std;

// } Driver Code Ends
//User function Template for C++
class Solution {
  public:
    string removeReverse(string S) {
        // code here
            string s = S;
        int freq[26] = {0};
        int flag = 0;
        int st=0; int end=s.length()-1; 
        
        // Count frequency of each character
        for(int i=0; i<s.length(); i++){
            char ch = s[i];
            freq[ch-'a']++;
        }
        
        while(st<=end) {
            //checking from start
            if(flag==0) {
                if(freq[s[st]-'a']==1) { // no repeating character, check for others
                    st++;
                } else {
                    freq[s[st]-'a']--;
                    s[st] = '#';
                    st++;
                    flag = 1; // string is reversed, traverse from 'end' ptr for flag=1
                }
            } else {
                if(freq[s[end]-'a']==1) { // no repeating character, check for others
                    end--;
                } else {
                    freq[s[end]-'a']--;
                    s[end] = '#';
                    end--;
                    flag = 0; // string is reversed, traverse from start
                }
            }
        }
        
        //after processing for flag==0, it passes to flag==1 (but here i cannot find any character freq >1)
        //but still, I have to reverse it
        if(flag==1) {
            reverse(s.begin(), s.end());
        }
        
        string ans="";
        for(int i=0; i<s.length(); i++) {
            if(s[i]!='#') {
                ans+=s[i];
            }
        }
        return ans;
    }
};

//{ Driver Code Starts.
int main() {
    int t;
    cin >> t;
    while (t--) {
        string S;
        cin>>S;
        Solution obj;
        auto ans = obj.removeReverse(S);
        cout<<ans<<endl;
    }
}
// } Driver Code Ends